pymc_experimental/inference/pathfinder/pathfinder.py

```diff
@@ -0,0 +1,980 @@
+    logdet += pt.sum(pt.log(alpha), axis=-1)
+
+    # NOTE: changed the sign from "x + " to "x -" of the expression to match Stan which differs from Zhang et al., (2022). same for dense version.
+    mu = x - (
+        # (L, N), (L, N) -> (L, N)
+        pt.batched_dot(alpha_diag, g)
+        # beta @ gamma @ beta.T
+        # (L, N, 2J), (L, 2J, 2J), (L, 2J, N) -> (L, N, N)
+        # (L, N, N), (L, N) -> (L, N)
+        + pt.batched_dot((beta @ gamma @ pt.matrix_transpose(beta)), g)
+    )
+
+    phi = pt.matrix_transpose(
+        # (L, N, 1)
+        mu[..., None]
+        # (L, N, N), (L, N, M) -> (L, N, M)
+        + sqrt_alpha_diag
+        @ (
+            # (L, N, 2J), (L, 2J, M) -> (L, N, M)
+            # intermediate calcs below
+            # (L, N, 2J), (L, 2J, 2J) -> (L, N, 2J)
+            (Q @ (Lchol - IdN))
+            # (L, 2J, N), (L, N, M) -> (L, 2J, M)
+            @ (pt.matrix_transpose(Q) @ pt.matrix_transpose(u))
+            # (L, N, M)
+            + pt.matrix_transpose(u)
+        )
+    )  # fmt: off
+
+    return phi, logdet
+
+
+def bfgs_sample(
+    rng,
+    num_samples: int,
+    x,  # position
+    g,  # grad
+    alpha,
+    beta,
+    gamma,
+    index: int | None = None,
+    # random_seed: RandomSeed | None = None,
+    # rng,
+):
+    # batch: L = 8
+    # alpha_l: (N,)         => (L, N)
+    # beta_l: (N, 2J)       => (L, N, 2J)
+    # gamma_l: (2J, 2J)     => (L, 2J, 2J)
+    # Q : (N, 2J)           => (L, N, 2J)
+    # R: (2J, 2J)           => (L, 2J, 2J)
+    # u: (M, N)             => (L, M, N)
+    # phi: (M, N)           => (L, M, N)
+    # logQ_phi: (M,)        => (L, M)
+    # Lchol: (2J, 2J)       => (L, 2J, 2J)
+    # theta: (J, N)
+
+    if index is not None:
+        x = x[index][None, ...]
+        g = g[index][None, ...]
+        alpha = alpha[index][None, ...]
+        beta = beta[index][None, ...]
+        gamma = gamma[index][None, ...]
+
+    L, N, JJ = beta.shape
+
+    (alpha_diag, inv_sqrt_alpha_diag, sqrt_alpha_diag), _ = pytensor.scan(
+        lambda a: [pt.diag(a), pt.diag(pt.sqrt(1.0 / a)), pt.diag(pt.sqrt(a))],
+        sequences=[alpha],
+        allow_gc=False,
+    )
+
+    u = pt.random.normal(size=(L, num_samples, N), rng=rng)
+
+    sample_inputs = (
+        x,
+        g,
+        alpha,
+        beta,
+        gamma,
+        alpha_diag,
+        inv_sqrt_alpha_diag,
+        sqrt_alpha_diag,
+        u,
+    )
+
+    phi, logdet = pytensor.ifelse(
+        JJ >= N,
+        bfgs_sample_dense(*sample_inputs),
+        bfgs_sample_sparse(*sample_inputs),
+    )
+
+    logQ_phi = -0.5 * (
+        logdet[..., None]
+        + pt.sum(u * u, axis=-1)
+        + N * pt.log(2.0 * pt.pi)
+    )  # fmt: off
+
+    mask = pt.isnan(logQ_phi) | pt.isinf(logQ_phi)
+    logQ_phi = pt.set_subtensor(logQ_phi[mask], pt.inf)
+    # phi: (L, M, N)
+    # logQ_phi: (L, M)
+    return phi, logQ_phi
+
+
+class LogLike(Op):
```

Comment:
Why is this an Op? How is it constructed? Why can't you work directly with PyTensor graphs?

Reply:
you can use `pt.vectorize/pytensor.graph.replace.vectorize_graph``